File: crates/engine/src/tools/web_search.rs
use super::web_cache;
use super::web_shared::next_user_agent;
use super::{str_arg, Tool, ToolContext, ToolFuture, ToolResult};
use serde_json::Value;
use std::collections::HashMap;
use std::time::Duration;

pub struct WebSearchTool;

impl Tool for WebSearchTool {
    fn name(&self) -> &str {
        "web_search"
    }

    fn description(&self) -> &str {
        "Search the web using DuckDuckGo. Returns a list of results with titles, URLs, and descriptions."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "The search query"
                }
            },
            "required": ["query"]
        })
    }

    fn needs_confirm(&self, args: &HashMap<String, Value>) -> Option<String> {
        let query = str_arg(args, "query");
        Some(query.to_string())
    }

    fn execute<'a>(
        &'a self,
        args: HashMap<String, Value>,
        _ctx: &'a ToolContext<'a>,
    ) -> ToolFuture<'a> {
        Box::pin(async move { tokio::task::block_in_place(|| run_search(&args)) })
    }
}

fn run_search(args: &HashMap<String, Value>) -> ToolResult {
    let query = str_arg(args, "query");
    if query.is_empty() {
        return ToolResult {
            content: "Query cannot be empty".into(),
            is_error: true,
        };
    }

    let cache_key = format!("search:{query}");
    if let Some(cached) = web_cache::get(&cache_key) {
        return ToolResult {
            content: cached,
            is_error: false,
        };
    }

    let search_query = query.clone();
    let fetch_result = std::thread::spawn(move || {
        let ua = next_user_agent();
        let client = reqwest::blocking::Client::builder()
            .timeout(Duration::from_secs(20))
            .user_agent(ua)
            .redirect(reqwest::redirect::Policy::limited(10))
            .build()?;

        let body = url::form_urlencoded::Serializer::new(String::new())
            .append_pair("q", &search_query)
            .append_pair("kl", "us-en")
            .finish();

        let response = client
            .post("https://html.duckduckgo.com/html/")
            .header("Content-Type", "application/x-www-form-urlencoded")
            .header("Accept", "text/html")
            .header("Accept-Language", "en-US,en;q=0.9")
            .header("Referer", "https://html.duckduckgo.com/html/")
            .header("Origin", "https://html.duckduckgo.com")
            .body(body)
            .send()?;

        if !response.status().is_success() {
            return Err(response.error_for_status().unwrap_err());
        }

        response.text()
    })
    .join();

    let html = match fetch_result {
        Ok(Ok(h)) => h,
        Ok(Err(e)) => {
            return ToolResult {
                content: format!("Search failed: {e}"),
                is_error: true,
            }
        }
        Err(_) => {
            return ToolResult {
                content: "Search thread panicked".into(),
                is_error: true,
            }
        }
    };

    let results = parse_ddg_results(&html);
    if results.is_empty() {
        return ToolResult {
            content: "No results found".into(),
            is_error: false,
        };
    }

    let mut output = String::new();
    for (i, r) in results.iter().enumerate() {
        output.push_str(&format!("{}. {}\n   {}\n", i + 1, r.title, r.link));
        if !r.description.is_empty() {
            output.push_str(&format!("   {}\n", r.description));
        }
        output.push('\n');
    }

    let output = output.trim_end().to_string();
    web_cache::put(&cache_key, &output);

    ToolResult {
        content: output,
        is_error: false,
    }
}

struct SearchResult {
    title: String,
    link: String,
    description: String,
}

fn parse_ddg_results(html: &str) -> Vec<SearchResult> {
    use scraper::{Html, Selector};

    let doc = Html::parse_document(html);
    let result_sel = Selector::parse("div.result, div.web-result").unwrap();
    let title_sel = Selector::parse("a.result__a").unwrap();
    let snippet_sel = Selector::parse("a.result__snippet").unwrap();

    let mut results = Vec::new();

    for el in doc.select(&result_sel) {
        if results.len() >= 20 {
            break;
        }

        let Some(title_el) = el.select(&title_sel).next() else {
            continue;
        };

        let title: String = title_el.text().collect::<String>().trim().to_string();
        if title.is_empty() {
            continue;
        }

        let raw_href = title_el.value().attr("href").unwrap_or("").to_string();
        let link = extract_ddg_url(&raw_href);
        if link.is_empty() {
            continue;
        }

        let description = el
            .select(&snippet_sel)
            .next()
            .map(|s| s.text().collect::<String>().trim().to_string())
            .unwrap_or_default();

        results.push(SearchResult {
            title,
            link,
            description,
        });
    }

    results
}

fn extract_ddg_url(ddg_url: &str) -> String {
    if ddg_url.contains("uddg=") {
        if let Some(start) = ddg_url.find("uddg=") {
            let after = &ddg_url[start + 5..];
            let encoded = if let Some(end) = after.find('&') {
                &after[..end]
            } else {
                after
            };
            return url::form_urlencoded::parse(encoded.as_bytes())
                .next()
                .map(|(k, v)| {
                    if v.is_empty() {
                        k.to_string()
                    } else {
                        format!("{k}={v}")
                    }
                })
                .unwrap_or_default();
        }
    }

    if ddg_url.starts_with("http://") || ddg_url.starts_with("https://") {
        return ddg_url.to_string();
    }

    String::new()
}

File: runtime/lua/smelt/tools/web_search.lua
-- Built-in web_search tool. DuckDuckGo HTML search with 15-minute cache and rotated UA.

local function urlencode(s)
  return (
    s:gsub("([^A-Za-z0-9_.~-])", function(c)
      return string.format("%%%02X", string.byte(c))
    end)
  )
end

smelt.tools.register({
  name = "web_search",
  description = "Search the web using DuckDuckGo. Returns a list of results with titles, URLs, and descriptions.",
  override = true,
  permission_defaults = { normal = "allow", plan = "allow", apply = "allow" },
  parameters = {
    type = "object",
    properties = {
      query = {
        type = "string",
        description = "The search query",
      },
    },
    required = { "query" },
  },
  summary = function(args)
    return args.query or ""
  end,
  render = function(args, output, ctx)
    return smelt.layout.text(output.content, {
      hl_group = output.is_error and "ErrorMsg" or nil,
    })
  end,
  execute = function(args)
    local query = args.query or ""
    if query == "" then
      return { content = "Query cannot be empty", is_error = true }
    end

    local cache_key = "search:" .. query
    local cached = smelt.http.cache.read(cache_key)
    if cached then
      return cached
    end

    local body = "q=" .. urlencode(query) .. "&kl=us-en"
    local resp, err = smelt.http.post("https://html.duckduckgo.com/html/", body, {
      timeout_secs = 20,
      max_redirects = 10,
      headers = {
        ["User-Agent"] = smelt.http.random_user_agent(),
        ["Content-Type"] = "application/x-www-form-urlencoded",
        ["Accept"] = "text/html",
        ["Accept-Language"] = "en-US,en;q=0.9",
        ["Referer"] = "https://html.duckduckgo.com/html/",
        ["Origin"] = "https://html.duckduckgo.com",
      },
    })
    if err then
      return { content = "Search failed: " .. err, is_error = true }
    end
    if resp.status < 200 or resp.status >= 300 then
      return { content = "Search failed: HTTP " .. resp.status, is_error = true }
    end

    local results = smelt.html.parse_ddg_results(resp.body)
    if #results == 0 then
      return "No results found"
    end

    local lines = {}
    for i, r in ipairs(results) do
      table.insert(lines, i .. ". " .. r.title)
      table.insert(lines, "   " .. r.link)
      if r.description and r.description ~= "" then
        table.insert(lines, "   " .. r.description)
      end
      table.insert(lines, "")
    end
    while #lines > 0 and lines[#lines] == "" do
      table.remove(lines)
    end

    local output = table.concat(lines, "\n")
    smelt.http.cache.write(cache_key, output)
    return output
  end,
})
